test_container/tests/test/java/java_modules.py

```diff
@@ -0,0 +1,90 @@
+class JavaModules(udf.TestCase):
+    env = None
+    java_udf_jar_java11 = None
+    java_udf_jar_java17 = None
+    
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.java_udf_jar_java11 = self.build_java_udf(11)
+        self.java_udf_jar_java17 = self.build_java_udf(17)
+
+    def setUp(self):
+        self.query('CREATE SCHEMA JAVA_MODULES', ignore_errors=True)
+        self.query('OPEN SCHEMA JAVA_MODULES')
+        self.env = docker_db_environment.DockerDBEnvironment("JAVA_MODULE_TEST")
+
+    def build_java_udf(self, javaVersion: int) -> Path:
+        result = subprocess.run(["mvn", "--batch-mode", "--file", java_udf_dir/"pom.xml", "clean", "package", f"-Djava.version={javaVersion}"], stdout=subprocess.PIPE)
+        result.check_returncode()
+        jar = java_udf_dir / f"target/test-udf-java-{javaVersion}.jar"
+        assert jar.exists()
+        target = Path(tempfile.gettempdir()) / f"test-udf-java-{javaVersion}.jar"
+        shutil.move(jar, target)
+        return target
+
+    def test_udf_jar_is_module(self):
+        with zipfile.ZipFile(self.java_udf_jar_java11, 'r') as zip:
+            assert "module-info.class" in zip.namelist()
+    
+    def test_udf_jar_is_module_java17(self):
+        with zipfile.ZipFile(self.java_udf_jar_java17, 'r') as zip:
+            assert "module-info.class" in zip.namelist()
+
+    def upload_to_bucketfs(self, path: Path) -> str:
+        docker_db_container = self.env.get_docker_db_container()
+        docker_db_ip = self.env.get_ip_address_of_container(docker_db_container)
+        upload_url = f"http://{docker_db_ip}:6583/myudfs/{path.name}"
+        username = "w"
+        password = "write"
+        r_upload = requests.put(upload_url, data=path.read_bytes(), auth=HTTPBasicAuth(username, password))
+        r_upload.raise_for_status()
+        return f"/buckets/bfsdefault/myudfs/{path.name}"
+
+
+    def test_module_jar_udf_classpath(self):
+        assert self.get_jre_version(self.java_udf_jar_java11) in ["11", "17"]
+
+    def test_java_17_udf(self):
+        if self.get_jre_version(self.java_udf_jar_java11) == "17":
+            bucketfs_path = self.upload_to_bucketfs(self.java_udf_jar_java17)
+            self.query(udf.fixindent(f'''
+                CREATE JAVA SCALAR SCRIPT JAVA_17_UDF() RETURNS INT AS
+                %scriptclass com.exasol.slc.testudf.Main;
+                %jar {bucketfs_path};
+                '''))
+            rows = self.query("SELECT JAVA_17_UDF()")
+            return str(rows[0][0])
+
+    def get_jre_version(self, jar: Path):
+        bucketfs_path = self.upload_to_bucketfs(jar)
+        self.query(udf.fixindent(f'''
+                CREATE JAVA SCALAR SCRIPT JRE_VERSION() RETURNS INT AS
+                %scriptclass com.exasol.slc.testudf.Main;
+                %jar {bucketfs_path};
+                '''))
+        rows = self.query("SELECT JRE_VERSION()")
+        return str(rows[0][0])
```

Comment:
I fear these tests need some explanation, because I wasn't sure for a moment, if they tested the same, but basically, we test here
- java 11 jar with java 11 udf
- java 11 jar with java 17 udf
- java 17 jar with java 17 udf

Reply:
I updated the tests. We now test all combinations of a Java 11&17 JAR running on JRE 11&17, expecting an error for 11 on 17.